File: java/0783.0_Minimum_Distance_Between_BST_Nodes.java
/*
DFS

Runtime: 0 ms, faster than 100.00% of Java online submissions for Minimum Distance Between BST Nodes.
Memory Usage: 40 MB, less than 58.93% of Java online submissions for Minimum Distance Between BST Nodes.
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int minDistance = Integer.MAX_VALUE;
    TreeNode prevValue;
    
    void inorderTraversal(TreeNode root) {
        if (root == null) {
            return;
        }
        inorderTraversal(root.left);
        if (prevValue != null) {
            minDistance = Math.min(minDistance, root.val - prevValue.val);
        }
        prevValue = root;
        inorderTraversal(root.right);
    }
        
    public int minDiffInBST(TreeNode root) {
        inorderTraversal(root);
        return minDistance;
    }
}

File: cpp/0783.0_Minimum_Distance_Between_BST_Nodes.cpp
/*
Runtime: 2 ms, faster than 69.56% of C++ online submissions for Minimum Distance Between BST Nodes.
Memory Usage: 9.7 MB, less than 69.86% of C++ online submissions for Minimum Distance Between BST Nodes.
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDistance = INT_MAX;
    TreeNode* prevValue;
    
    void inorderTraversal(TreeNode* root) {
        if (root == NULL) {
            return;
        }
        inorderTraversal(root->left);
        if (prevValue != NULL) {
            minDistance = min(minDistance, root->val - prevValue->val);
        }
        prevValue = root;
        inorderTraversal(root->right);
    }
    
    int minDiffInBST(TreeNode* root) {
        inorderTraversal(root);
        return minDistance;
    }
};
